### PR TITLE
explicitly removes accounts with zero lamports from stakes cache

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
Zero lamport accounts are not stored in accounts-db:
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/accounts_db.rs#L5007-L5014
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/accounts_db.rs#L4481-L4483

However, in order to purge the account from cache, stakes cache update
partially relies on accounts data to be also zeroed out :
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/bank.rs#L3471-L3478

This can be error-prone and introduce inconsistency between accounts-db
and vote/stake cache. This commit instead explicitly removes accounts
from cache if `lamports == 0`.

#### Summary of Changes
explicitly remove accounts with zero lamports from stakes cache

### CODE REVIEW

File: runtime/src/stakes.rs [hunk 1]
```diff
@@ -62,28 +62,44 @@ impl StakesCache {
         // but the owner changes, then this needs to evict the account from
         // the cache. see:
         // https://github.com/solana-labs/solana/pull/24200#discussion_r849935444
-        if solana_vote_program::check_id(account.owner()) {
-            let new_vote_account = if account.lamports() != 0
-                && VoteState::is_correct_size_and_initialized(account.data())
-            {
+        let owner = account.owner();
+        // Zero lambort accounts are not stored in accounts-db
```
Comment (jstarry) [inline runtime/src/stakes.rs:66]:
```suggestion
        // Zero lamport accounts are not stored in accounts-db
```

File: runtime/src/stakes.rs [hunk 2]
```diff
@@ -62,28 +62,44 @@ impl StakesCache {
...
+        // and so should be removed from cache as well.
+        if account.lamports() == 0 {
+            if solana_vote_program::check_id(owner) {
+                let mut stakes = self.0.write().unwrap();
+                stakes.remove_vote_account(pubkey);
+            } else if solana_stake_program::check_id(owner) {
```
Comment (jstarry) [inline runtime/src/stakes.rs:72]: Before your change, `update_stake_delegation` would still store an initialized delegation even if it had zero lamports. However, the stake program always sets state to uninitialized when it drains a stake account to zero lamports. It also doesn't allow non-rent-exempt stake accounts to be created. So this looks safe to me without a feature gate.

**Review by jstarry — APPROVED** [review]

File: runtime/src/stakes.rs (hunk 1 above)
Comment (behzadnouri) [reply runtime/src/stakes.rs]: fixed, thanks

File: runtime/src/stakes.rs (hunk 2 above)
Comment (behzadnouri) [reply runtime/src/stakes.rs:72]: yup, that is consistent with my thinking.
Are there any places then that rely on this logic?
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/bank.rs#L3471-L3478

Comment (jstarry) [reply runtime/src/stakes.rs:72]: I'm not really sure what you're asking. Which logic are you worried that we're relying on in other places?

File: runtime/src/vote_account.rs [hunk 3]
```diff
@@ -125,18 +125,18 @@ impl VoteAccounts {
         entry
     }
 
-    pub(crate) fn add_stake(&mut self, pubkey: &Pubkey, delta: u64) {
+    pub(crate) fn add_stake(&mut self, voter_pubkey: &Pubkey, delta: u64) {
```
[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/vote_account.rs]: Yeah, that's true. Fine with me

File: runtime/src/stakes.rs (hunk 2 above)
Comment (behzadnouri) [reply runtime/src/stakes.rs:72]: I mean here:
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/bank.rs#L3471-L3478
it is zeroing out the data in addition to `lamports = 0`.

Are there any places in the code that distinguish between these two accounts:
* an account with `lamports == 0` and `data == [0; size]`
* an account with `lamports == 0` but non zero data;

Comment (jstarry) [reply runtime/src/stakes.rs:72]: Oh I see, good point. Another place would be eager rent collection which just sets lamports to 0.  Can't think of any others

**Pull request merged**